exporter/trace/stackdriver/trace.go

```diff
@@ -0,0 +1,182 @@
+	ctx := o.Context
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	client, err := traceclient.NewClient(ctx, o.TraceClientOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("Stackdriver: couldn't initiate trace client: %v", err)
+	}
+	return newTraceExporterWithClient(o, client), nil
+}
+
+const defaultBufferedByteLimit = 0 * 1024 * 1024
+
+func newTraceExporterWithClient(o *options, c *traceclient.Client) *traceExporter {
+	e := &traceExporter{
+		projectID: o.ProjectID,
+		client:    c,
+		o:         o,
+	}
+	b := bundler.NewBundler((*tracepb.Span)(nil), func(bundle interface{}) {
+		e.uploadFn(bundle.([]*tracepb.Span))
+	})
+	if o.BundleDelayThreshold > 0 {
+		b.DelayThreshold = o.BundleDelayThreshold
+	} else {
+		b.DelayThreshold = 2 * time.Second
+	}
+	if o.BundleCountThreshold > 0 {
+		b.BundleCountThreshold = o.BundleCountThreshold
+	} else {
+		b.BundleCountThreshold = 50
+	}
+	// The measured "bytes" are not really bytes, see exportReceiver.
+	b.BundleByteThreshold = b.BundleCountThreshold * 200
+	b.BundleByteLimit = b.BundleCountThreshold * 1000
+	if o.TraceSpansBufferMaxBytes > 0 {
+		b.BundleByteLimit = o.TraceSpansBufferMaxBytes
+	} else {
+		b.BundleByteLimit = defaultBufferedByteLimit
+	}
+
+	e.bundler = b
+	e.uploadFn = e.uploadSpans
+	return e
+}
+
+// ExportSpan exports a SpanData to Stackdriver Trace.
+func (e *traceExporter) ExportSpan(s *trace.SpanData) {
+	protoSpan := protoFromSpanData(s, e.projectID)
+	protoSize := proto.Size(protoSpan)
+	err := e.bundler.Add(protoSpan, protoSize)
+	switch err {
+	case nil:
+		return
+	case bundler.ErrOversizedItem:
+	case bundler.ErrOverflow:
+		e.overflowLogger.log()
+	default:
+		e.o.handleError(err)
+	}
+}
+
+// Flush waits for exported trace spans to be uploaded.
+//
+// This is useful if your program is ending and you do not want to lose recent
+// spans.
+func (e *traceExporter) Flush() {
+	e.bundler.Flush()
+}
+
+// uploadSpans uploads a set of spans to Stackdriver.
+func (e *traceExporter) uploadSpans(spans []*tracepb.Span) {
+	req := tracepb.BatchWriteSpansRequest{
+		Name:  "projects/" + e.projectID,
+		Spans: spans,
+	}
+	// Create a never-sampled span to prevent traces associated with exporter.
+	ctx, cancel := newContextWithTimeout(e.o.Context, e.o.Timeout)
+	defer cancel()
+
+	// TODO(ymotongpoo): add this part after OTel support NeverSampler
+	// for tracer.Start() initialization.
+	//
+	// tracer := apitrace.Register()
+	// ctx, span := tracer.Start(
+	// 	ctx,
+	// 	"go.opentelemetry.io/exporter/stackdriver.uploadSpans",
+	// )
+	// defer span.End()
+	// span.SetAttribute(key.New("num_spans").Int64(int64(len(spans))))
+
+	err := e.client.BatchWriteSpans(ctx, &req)
+	if err != nil {
+		// TODO(ymotongpoo): handle detailed error categories
+		// span.SetStatus(codes.Unknown)
+		e.o.handleError(err)
+	}
+}
+
+// overflowLogger ensures that at most one overflow error log message is
+// written every 5 seconds.
+type overflowLogger struct {
```

Comment:
I believe we've discussed handling errors of this nature in a more holistic manner. Every bit of code needs to pass errors back to the user, and not all users will want these on the console. We have discussed registering a callback with the SDK as a means of conveying errors.

Reply:
Eventually, overflowLogger was totally removed as a result of elimination of bundler.